Classes/common/HTTP/CDTURLSessionInterceptorContext.h

```diff
@@ -0,0 +1,39 @@
+//
+//  CDTURLSEssionFilterContext.h
+//  
+//
+//  Created by Rhys Short on 17/08/2015.
+//
+//
+
+#import <Foundation/Foundation.h>
+#import "CDTMacros.h"
+
+@interface CDTURLSessionInterceptorContext : NSObject
+
+@property (nonnull, readonly, nonatomic, strong) NSMutableURLRequest *request;
+
+- (nullable instancetype) init UNAVAILABLE_ATTRIBUTE;
+
+/**
+ *  Initalizes a CDTURLSessionInterceptorContext
+ *
+ *  @param request the request this context should represent
+ *
+ **/
+- (nullable instancetype) initWithRequest:(nonnull NSMutableURLRequest*)request NS_DESIGNATED_INITIALIZER;
+
+
+@end
+
+@protocol CDTURLSessionInterceptor <NSObject>
+
+/**
+ *  Intercepts a request before it is made
+ *  @param context the context for this interception
+ * 
+ *  @return the context for this interception
+ **/
+@optional - (nonnull CDTURLSessionInterceptorContext*) interceptRequestInContext:(nonnull CDTURLSessionInterceptorContext*)context;
```

Comment:
I plan on adding the `interceptResponse:` method in this protocol too. By making them optional it means the user can implement either, or both.